tests/scripts/generate_code.py

```diff
@@ -0,0 +1,710 @@
+    return out
+
+
+def parse_test_data(data_f, debug=False):
+    """
+    Parses .data file
+    
+    :param data_f: 
+    :return: 
+    """
+    STATE_READ_NAME = 0
+    STATE_READ_ARGS = 1
+    state = STATE_READ_NAME
+    deps = []
+    name = ''
+    for line in data_f:
+        line = line.strip()
+
+        # Blank line indicates end of test
+        if len(line) == 0:
+            assert state != STATE_READ_ARGS, "Newline before arguments. " \
+                                                 "Test function and arguments missing for %s" % name
+            continue
+
+        if state == STATE_READ_NAME:
+            # Read test name
+            name = line
+            state = STATE_READ_ARGS
+        elif state == STATE_READ_ARGS:
+            # Check dependencies
+            m = re.search('depends_on\:(.*)', line)
+            if m:
+                deps = [x.strip() for x in m.group(1).split(':') if len(x.strip())]
+            else:
+                # Read test vectors
+                parts = escaped_split(line, ':')
+                function = parts[0]
+                args = parts[1:]
+                yield name, function, deps, args
+                deps = []
+                state = STATE_READ_NAME
+    assert state != STATE_READ_ARGS, "Newline before arguments. " \
+                                     "Test function and arguments missing for %s" % name
+
+
+def gen_dep_check(dep_id, dep):
+    """
+    Generate code for the dependency.
+    
+    :param dep_id: 
+    :param dep: 
+    :return: 
+    """
+    assert dep_id > -1, "Dependency Id should be a positive integer."
+    noT, dep = ('!', dep[1:]) if dep[0] == '!' else ('', dep)
+    assert len(dep) > 0, "Dependency should not be an empty string."
+    dep_check = '''
+        case {id}:
+            {{
+#if {noT}defined({macro})
+                ret = DEPENDENCY_SUPPORTED;
+#else
+                ret = DEPENDENCY_NOT_SUPPORTED;
+#endif
+            }}
+            break;'''.format(noT=noT, macro=dep, id=dep_id)
+    return dep_check
+
+
+def gen_expression_check(exp_id, exp):
+    """
+    Generates code for expression check
+    
+    :param exp_id: 
+    :param exp: 
+    :return: 
+    """
+    assert exp_id > -1, "Expression Id should be a positive integer."
+    assert len(exp) > 0, "Expression should not be an empty string."
+    exp_code = '''
+        case {exp_id}:
+            {{
+                *out_value = {expression};
+            }}
+            break;'''.format(exp_id=exp_id, expression=exp)
+    return exp_code
+
+
+def write_deps(out_data_f, test_deps, unique_deps):
+    """
+    Write dependencies to intermediate test data file.
+    It also returns dependency check code.
+      
+    :param out_data_f:
+    :param dep: 
+    :param unique_deps: 
+    :return: 
+    """
+    dep_check_code = ''
+    if len(test_deps):
+        out_data_f.write('depends_on')
+        for dep in test_deps:
+            if dep not in unique_deps:
```

Comment:
I can implement it like this:
```
            dep_id = unique_deps.index(dep) if dep in unique_deps else -1
            if dep == -1:
                dep_id = len(unique_deps)
                unique_deps.append(dep)
                dep_check_code += gen_dep_check(dep_id, dep)
```
But then there are two ```if``` check. Performance is not a concern here but have ```-1``` for another condition looks useless.